server: Enforce Session Authentication (Optional)

mrtracy opened the pull request:
Enforces session authentication by wrapping the Admin, Status and
TimeSeries handlers inside of an authenticating handler which will
reject requests without a valid session cookie.

Session enforcement is *off* for new Cockroach clusters, and there is no
command line option to enable it (as there is no UI support to actually
create a session).

However, authentication is now *on* by default for TestServer. All tests
which access authenticated APIs have either been modified to use an
authenticated client (with a valid session cookie), or explicitly turn
off session enforcement with a new test server option.

Commit:
server: Enforce Session Authentication (Optional)

Enforces session authentication by wrapping the Admin, Status and
TimeSeries handlers inside of an authenticating handler which will
reject requests without a valid session cookie.

Session enforcement is *off* for new Cockroach clusters, and there is no
command line option to enable it (as there is no UI support to actually
create a session).

However, authentication is now *on* by default for TestServer. All tests
which access authenticated APIs have either been modified to use an
authenticated client (with a valid session cookie), or explicitly turn
off session enforcement with a new test server option.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17499)
<!-- Reviewable:end -->

tamird commented:
Reviewed 11 of 12 files at r1.
Review status: 11 of 12 files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[pkg/security/certs_rotation_test.go, line 60 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCga9eXL2n3I12LHF:-KqzCga9eXL2n3I12LHG:b-kvec41) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/security/certs_rotation_test.go#L60)):*
> ```Go
> 
> 	// Start a test server with first set of certs.
> 	// Disable web session auth for simplicity.
> ```

isn't it just as simple to use an authenticated client? looks like just a few lines would change.

---

*[pkg/security/certs_test.go, line 215 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCmCV3Ix6VuK6ugG_:-KqzCmCV3Ix6VuK6ugGa:bq99osr) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/security/certs_test.go#L215)):*
> ```Go
> 	// Start a test server and override certs.
> 	// We use a real context since we want generated certs.
> 	// Disable web session authentication for simplicity.
> ```

ditto - what's the big simplification here?

---

*[pkg/server/admin_test.go, line 1064 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzArdVtxrg9ujZJ2-z:-KqzArdVtxrg9ujZJ20-:b-ivee6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/admin_test.go#L1064)):*
> ```Go
> 
> 	// Health API is not accessible if the node is not accessible, because it
> 	// cannot verify the authentication session.
> ```

even the health API requires authentication? is this wise?

---

*[pkg/server/authentication.go, line 314 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzAzU1tUU0L5yqKUP3:-KqzAzU1tUU0L5yqKUP4:bkd1o74) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication.go#L314)):*
> ```Go
> }
> 
> // ServeHTTP implements http.Handler.
> ```

this comment is not necessary because `authenticationMux` is not exported.

---

*[pkg/server/authentication.go, line 318 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzBGxiYwYReQvcIgmc:-KqzBGxiYwYReQvcIgmd:b-3c0hnc) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication.go#L318)):*
> ```Go
> 	// Validate the returned cookie.
> 	rawCookie, err := req.Cookie(sessionCookieName)
> 	if err != nil {
> ```

is it wise to discard the error completely? throughout.

---

*[pkg/server/authentication.go, line 365 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzB5ySvutZZZKBv-hK:-KqzB5ySvutZZZKBv-hL:bjto5gc) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication.go#L365)):*
> ```Go
> 	cookieBytes, err := base64.StdEncoding.DecodeString(encodedCookie.Value)
> 	if err != nil {
> 		return nil, fmt.Errorf("session cookie could not be decoded: %s", err)
> ```

errors.Wrap? throughout.

---

*[pkg/server/authentication_test.go, line 72 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCqK0XRPpaoSiA5rr:-KqzCqK0XRPpaoSiA5rs:b-wbu607) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L72)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 	s, _, _ := serverutils.StartServer(t, base.TestServerArgs{
> 		// This test is verifying the (unimplemented) authentication of SSL
> ```

is there anything meaningful being saved by disabling it?

---

*[pkg/server/authentication_test.go, line 337 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzD4_x4zf6wLVLuP3S:-KqzD4_x4zf6wLVLuP3T:b-rq1zcg) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L337)):*
> ```Go
> 		{
> 			cookie: serverpb.SessionCookie{
> 				Id:     id,
> ```

nit: give this a custom name so it can be "ID"?

---

*[pkg/server/authentication_test.go, line 372 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzBVj2ILMzIA-kMt6u:-KqzBVj2ILMzIA-kMt6v:b-tqbnmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L372)):*
> ```Go
> 			shouldVerify: false,
> 		},
> 		// Empty.
> ```

instead of comments, can you give this struct a name field and then use that in the call to `t.Run`? that would make for better error messages in these tests.

---

*[pkg/server/authentication_test.go, line 434 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzDhCk7q1aYvjmevy_:-KqzDhCk7q1aYvjmevya:b-cbs90m) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L434)):*
> ```Go
> 		t.Fatalf("login got error %s, wanted error with 401 status", err)
> 	}
> 	if e := headers.Get("Set-Cookie"); e != "" {
> ```

i missed this in the earlier review, but why not return the `http.Response` from `tryLogin` and just call `Cookies` on it? http://godoc.org/net/http#Response.Cookies

---

*[pkg/server/authentication_test.go, line 540 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzE5S-VXG4WGexE9OW:-KqzE5S032feWgag8bsn:bsz4h2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L540)):*
> ```Go
> 	}
> 
> 	checkIsBehindAuth := func(method string, path string, body []byte) {
> ```

why this strange style rather than a table driven test?

---

*[pkg/server/server.go, line 964 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzBmXDhc0apWW9CSjB:-KqzBmXDhc0apWW9CSjC:buj58ky) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/server.go#L964)):*
> ```Go
> 	s.stopper.AddCloser(stop.CloserFn(gwCancel))
> 
> 	var authHandler http.Handler
> ```

perhaps:
```
authHandler := gwMux
if s.cfg.RequireWebSession() {
  authHandler = newAuthenticationMux(s.authentication, authHandler)
}
```

---

*[pkg/server/testserver.go, line 427 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzEM-4ZL4bGuVJr2q6:-KqzEM-4ZL4bGuVJr2q7:bk3f015) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/testserver.go#L427)):*
> ```Go
> 		// currently have an authorization mechanism, so a specific user is not
> 		// necessary.
> 		id, secret, err := ts.authentication.newAuthSession(context.TODO(), "authentic_user")
> ```

this whole thing might read a bit easier as:

```
ts.authClient.err =  func() error {
  id, secret, err := ...
  if err != nil {
    return err
  }
}()
```
since it doesn't require careful pairing of `err` assignments and `return`s.

---

*[pkg/testutils/serverutils/test_server_shim.go, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCXVg65ZBrJeRb1_V:-KqzCXVg65ZBrJeRb1_W:b7xvcdj) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/testutils/serverutils/test_server_shim.go#L104)):*
> ```Go
> 	// AdminURL returns the URL for the admin UI.
> 	AdminURL() string
> 	// GetHTTPClient returns an http client configured with the context's client
> ```

what context?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 4 of 15 files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/security/certs_rotation_test.go, line 60 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCga9eXL2n3I12LHF:-Kr7Row2VLh9RpGZ-2WI:boyyam) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/security/certs_rotation_test.go#L60)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

isn't it just as simple to use an authenticated client? looks like just a few lines would change.
</blockquote></details>

I would have to create an abstraction of "GetAuthenticatedHTTPClient", because this test does not get its HttpClients from the TestServer, but rather generates them from several different custom contexts.

This test is focused on certificate authentication, not cookie authentication, and it was much easier to disable auth for this test.

---

*[pkg/security/certs_test.go, line 215 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCmCV3Ix6VuK6ugG_:-Kr7SV88hxcmAaegppFX:b-n02alz) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/security/certs_test.go#L215)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto - what's the big simplification here?
</blockquote></details>

Same as above, its more work and not necessary.

---

*[pkg/server/admin_test.go, line 1064 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzArdVtxrg9ujZJ2-z:-Kr7Sb0GjWUmCy9SSnpW:b-1sck0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/admin_test.go#L1064)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

even the health API requires authentication? is this wise?
</blockquote></details>

I'm throwing everything behind authentication and we can move individual methods out as needed.  Excluding a method from authentication will require moving it to a different GRPC service.

---

*[pkg/server/authentication.go, line 314 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzAzU1tUU0L5yqKUP3:-Kr7SvLm1o_iZbLUIUs_:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication.go#L314)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this comment is not necessary because `authenticationMux` is not exported.
</blockquote></details>

Removed.

---

*[pkg/server/authentication.go, line 318 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzBGxiYwYReQvcIgmc:-KrCXpvXN8PoxWyMISip:brufwcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication.go#L318)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is it wise to discard the error completely? throughout.
</blockquote></details>

After taking another look, I believe we need to rethink the way we're handling internal server errors.

My main reason for treating this differently is this issue: https://www.owasp.org/index.php/Improper_Error_Handling

Essentially, the idea that an attacker can discover information about your system through subtle information leaks in edge cases such as errors.  We are an open source system, so the basic product has no such concern with leaking information about code; however, if a customer is running some sort of proprietary modification we would want to avoid leaking that information.  We *will* need to prevent information about the *data* on a cluster from leaking through error messages.

In this special case (the authentication mux), we are already the top level HTTP handler, so we should be explicitly careful about returning information to the client.  In the first two cases (cookie decoding problems), I think it's fine to return information about the specific error to the client and I've modified it as such; however, my instinct tells me that we shouldn't return an error if it came from the SQL layer, so I'm going to use `apiInternalError` in that case; this logs the error text to the server log, and returns a generic "internal server error" message to the client.

---

*[pkg/server/authentication.go, line 365 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzB5ySvutZZZKBv-hK:-KrC_w7QBfB018KDFQNT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication.go#L365)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

errors.Wrap? throughout.
</blockquote></details>

Done.

---

*[pkg/server/authentication_test.go, line 72 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCqK0XRPpaoSiA5rr:-KrCaDmXnJP32R41X6lz:b-q64pkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L72)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is there anything meaningful being saved by disabling it?
</blockquote></details>

See above.

---

*[pkg/server/authentication_test.go, line 337 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzD4_x4zf6wLVLuP3S:-KrD2H4t2scTpFccMsx7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L337)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: give this a custom name so it can be "ID"?
</blockquote></details>

Done.

---

*[pkg/server/authentication_test.go, line 372 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzBVj2ILMzIA-kMt6u:-KrCkiPwUcHQnB7ON-jj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L372)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

instead of comments, can you give this struct a name field and then use that in the call to `t.Run`? that would make for better error messages in these tests.
</blockquote></details>

Done.

---

*[pkg/server/authentication_test.go, line 434 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzDhCk7q1aYvjmevy_:-KrCf1__p0MNaHZCTrDJ:b-f3jcgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L434)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i missed this in the earlier review, but why not return the `http.Response` from `tryLogin` and just call `Cookies` on it? http://godoc.org/net/http#Response.Cookies
</blockquote></details>

Because it is related deeply to the factoring of the `httputil.{Post|Get}JSON` methods, which originally took an HttpClient and abstracted away *everything* else, dealing entirely in proto messages.

I already hacked together that PostJSONWithHeaders bit; I suppose I could change it to `PostJSONWithResponse`or something, since this is the only place I'm using it.  For now, I think it is relatively equivalent and I'm just going to leave it as is.

A possible long term solution would be to convert the `PostJSON/GetJSON` methods into a struct that wraps an http client, which can be outfitted with additional state while keeping the common case simple.

---

*[pkg/server/authentication_test.go, line 540 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzE5S-VXG4WGexE9OW:-KrCk2aZJQ_42UMhMA-D:b-njy2r4) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L540)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why this strange style rather than a table driven test?
</blockquote></details>

It was originally a table driven test, but my initial factoring was frustrating in the case of the POST. However, after looking at the resulting code, the `checkIsBehindAuth` method can easily be replaced with at table.  Done.

---

*[pkg/server/server.go, line 964 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzBmXDhc0apWW9CSjB:-KrCcKEK45R7m7m8zH8W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/server.go#L964)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

perhaps:
```
authHandler := gwMux
if s.cfg.RequireWebSession() {
  authHandler = newAuthenticationMux(s.authentication, authHandler)
}
```
</blockquote></details>

Done.

---

*[pkg/server/testserver.go, line 427 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzEM-4ZL4bGuVJr2q6:-KrCcAOuqFvLYGGEZhAV:b-83v4qp) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/testserver.go#L427)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this whole thing might read a bit easier as:

```
ts.authClient.err =  func() error {
  id, secret, err := ...
  if err != nil {
    return err
  }
}()
```
since it doesn't require careful pairing of `err` assignments and `return`s.
</blockquote></details>

Neat.

---

*[pkg/testutils/serverutils/test_server_shim.go, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCXVg65ZBrJeRb1_V:-KrCbOEywqwNflU5NvyR:ba01i1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/testutils/serverutils/test_server_shim.go#L104)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what context?
</blockquote></details>

Meant configuration.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 11 of 11 files at r2.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[pkg/security/certs_rotation_test.go, line 60 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCga9eXL2n3I12LHF:-KrD6Tm84MPdG0UFgElf:b4om8sk) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/security/certs_rotation_test.go#L60)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I would have to create an abstraction of "GetAuthenticatedHTTPClient", because this test does not get its HttpClients from the TestServer, but rather generates them from several different custom contexts.

This test is focused on certificate authentication, not cookie authentication, and it was much easier to disable auth for this test.
</blockquote></details>

I see; could you perhaps elaborate this in the comment?

---

*[pkg/security/certs_test.go, line 215 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCmCV3Ix6VuK6ugG_:-KrD6gNXXIA4Z_nRncbJ:b-74dv3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/security/certs_test.go#L215)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Same as above, its more work and not necessary.
</blockquote></details>

Ditto, an explanation of what work is being saved would be helpful to future (and current) readers.

---

*[pkg/server/admin_test.go, line 1064 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzArdVtxrg9ujZJ2-z:-KrD5S8KHJLaUhd4m6Zq:blio63z) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/admin_test.go#L1064)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I'm throwing everything behind authentication and we can move individual methods out as needed.  Excluding a method from authentication will require moving it to a different GRPC service.
</blockquote></details>

OK...won't this break our load balancers?

---

*[pkg/server/authentication.go, line 318 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzBGxiYwYReQvcIgmc:-KrD7xbQpY0uYA_1r0Dk:booh8qr) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication.go#L318)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

After taking another look, I believe we need to rethink the way we're handling internal server errors.

My main reason for treating this differently is this issue: https://www.owasp.org/index.php/Improper_Error_Handling

Essentially, the idea that an attacker can discover information about your system through subtle information leaks in edge cases such as errors.  We are an open source system, so the basic product has no such concern with leaking information about code; however, if a customer is running some sort of proprietary modification we would want to avoid leaking that information.  We *will* need to prevent information about the *data* on a cluster from leaking through error messages.

In this special case (the authentication mux), we are already the top level HTTP handler, so we should be explicitly careful about returning information to the client.  In the first two cases (cookie decoding problems), I think it's fine to return information about the specific error to the client and I've modified it as such; however, my instinct tells me that we shouldn't return an error if it came from the SQL layer, so I'm going to use `apiInternalError` in that case; this logs the error text to the server log, and returns a generic "internal server error" message to the client.
</blockquote></details>

That seems fine, but you're not using `apiInternalError` in this and the next error case.

---

*[pkg/server/authentication.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD6qTiifLGmMrbYXdu:-KrD6qTiifLGmMrbYXdv:bbwfic4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication.go#L189)):*
> ```Go
> 			datum[1].ResolvedType() != parser.TypeString ||
> 			datum[2].ResolvedType() != parser.TypeTimestamp {
> 			return fmt.Errorf("values returned from auth session lookup do not match expectation")
> ```

errors.Errorf, please.

---

*[pkg/server/authentication.go, line 197 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD7A0vOocPvoChhgSO:-KrD7A0vOocPvoChhgSP:b1snkbu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication.go#L197)):*
> ```Go
> 		username = string(*datum[1].(*parser.DString))
> 		expiresAt = datum[2].(*parser.DTimestamp).Time
> 		isRevoked = !(datum[3].ResolvedType() == parser.TypeNull)
> ```

why is this not `datum[3].ResolvedType() != parser.TypeNull`?

---

*[pkg/server/authentication.go, line 211 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD7JFKm0hnIJU9E1CY:-KrD7JFKm0hnIJU9E1CZ:b62x3iq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication.go#L211)):*
> ```Go
> 	}
> 
> 	now := s.server.clock.PhysicalTime()
> ```

nit: move into the if or remove altogether?

---

*[pkg/server/authentication.go, line 309 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD7aGqLF3uygphHY9w:-KrD7aGqLF3uygphHY9x:b-pky2hz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication.go#L309)):*
> ```Go
> }
> 
> func (am authenticationMux) ServeHTTP(w http.ResponseWriter, req *http.Request) {
> ```

probably better for this to have a pointer receiver, so that only the pointer type implements the interface. it's largely the same, but has better type safety because the erroneous case of downcasting a `http.Handler` to `authenticationMux` would be caught at compile time (as opposed to downcasting to `*authenticationMux`).

---

*[pkg/server/authentication.go, line 333 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD8IF1nFLiQrHVMm4R:-KrD8IF1nFLiQrHVMm4S:b-wjd4fv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication.go#L333)):*
> ```Go
> 	}
> 
> 	// TODO: At this point, we should set the session ID and username on the
> ```

put your name on this TODO

---

*[pkg/server/authentication.go, line 343 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD8P539RVshko0TAtB:-KrD8P549oFEDRA9u3Il:b96fhpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication.go#L343)):*
> ```Go
> 	cookieValueBytes, err := sessionCookie.Marshal()
> 	if err != nil {
> 		return nil, err
> ```

no wrap on this one? you have one in the decode method.

---

*[pkg/server/authentication_test.go, line 72 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCqK0XRPpaoSiA5rr:-KrD6aMEJh-xpO-JarSO:b-3jvtla) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L72)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

See above.
</blockquote></details>

As above, this comment explains why it is OK to disable authentication, but not why it is helpful to do so.

---

*[pkg/server/authentication_test.go, line 434 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzDhCk7q1aYvjmevy_:-KrD5kj2lOVi_3XblyUh:btmhguv) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L434)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Because it is related deeply to the factoring of the `httputil.{Post|Get}JSON` methods, which originally took an HttpClient and abstracted away *everything* else, dealing entirely in proto messages.

I already hacked together that PostJSONWithHeaders bit; I suppose I could change it to `PostJSONWithResponse`or something, since this is the only place I'm using it.  For now, I think it is relatively equivalent and I'm just going to leave it as is.

A possible long term solution would be to convert the `PostJSON/GetJSON` methods into a struct that wraps an http client, which can be outfitted with additional state while keeping the common case simple.
</blockquote></details>

Seems to me that `PostJSONWithHeaders` -> `PostJSONWithResponseor` would be an easy change to make and neatly address the current issues.

---

*[pkg/server/authentication_test.go, line 536 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD6vs248d6_J3B3VqO:-KrD6vs248d6_J3B3VqP:b-4utsdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication_test.go#L536)):*
> ```Go
> 				message = []byte(err.Error())
> 			}
> 			return fmt.Errorf("got status code %d (msg %s), wanted %d", a, string(message), e)
> ```

errors.Errorf, if you don't mind.

---

*[pkg/server/authentication_test.go, line 562 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD5zDUvHhJjbLgH_ly:-KrD5zDUvHhJjbLgH_lz:b66jyri) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication_test.go#L562)):*
> ```Go
> 		{"POST", ts.URLPrefix + "query", tsReqBuffer.Bytes()},
> 	} {
> 		t.Run("test auth for "+tc.path, func(t *testing.T) {
> ```

instead of "test auth for ", use a key=value pair, as some tools can use this as metadata:

```
t.Run(fmt.Sprintf("path=%s", tc.path), ...
```

---

*[pkg/server/server.go, line 981 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD6H6ZboVwTiQ01a-T:-KrD6H6ZboVwTiQ01a-U:b-n2suir) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/server.go#L981)):*
> ```Go
> 	var authHandler http.Handler = gwMux
> 	if s.cfg.RequireWebSession() {
> 		authHandler = newAuthenticationMux(s.authentication, gwMux)
> ```

s/gwMux/authHandler/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 9 of 15 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/security/certs_rotation_test.go, line 60 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCga9eXL2n3I12LHF:-KrU5qsg2FmzIlK0Pw7L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/security/certs_rotation_test.go#L60)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I see; could you perhaps elaborate this in the comment?
</blockquote></details>

Done.

---

*[pkg/security/certs_test.go, line 215 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCmCV3Ix6VuK6ugG_:-KrU5wlJRQih5WKGTrQU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/security/certs_test.go#L215)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Ditto, an explanation of what work is being saved would be helpful to future (and current) readers.
</blockquote></details>

Done.

---

*[pkg/server/admin_test.go, line 1064 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzArdVtxrg9ujZJ2-z:-KrU2xbJeU91kf6Q76-R:b-3xvmva) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/admin_test.go#L1064)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

OK...won't this break our load balancers?
</blockquote></details>

I assumed they used the `/health` endpoint, which I did not put behind authentication.  Even if the load balancers did depend on a non-authenticated endpoint, this commit wouldn't break them because it's not turned on.

---

*[pkg/server/authentication.go, line 318 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzBGxiYwYReQvcIgmc:-KrU2obZF4aQK_KaWJbm:bsocnqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication.go#L318)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

That seems fine, but you're not using `apiInternalError` in this and the next error case.
</blockquote></details>

It's not an internal error, it's an error with the request.

---

*[pkg/server/authentication.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD6qTiifLGmMrbYXdu:-KrU64m1UlCoAvNCCQD2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication.go#L189)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

errors.Errorf, please.
</blockquote></details>

Done.

---

*[pkg/server/authentication.go, line 197 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD7A0vOocPvoChhgSO:-KrU67AKpYZFIGfeLaq1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication.go#L197)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why is this not `datum[3].ResolvedType() != parser.TypeNull`?
</blockquote></details>

Done.

---

*[pkg/server/authentication.go, line 211 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD7JFKm0hnIJU9E1CY:-KrU6dcs5DxEv_GMeiQa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication.go#L211)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: move into the if or remove altogether?
</blockquote></details>

Done.

---

*[pkg/server/authentication.go, line 309 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD7aGqLF3uygphHY9w:-KrU6yc-KNmcF0MWfoF7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication.go#L309)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

probably better for this to have a pointer receiver, so that only the pointer type implements the interface. it's largely the same, but has better type safety because the erroneous case of downcasting a `http.Handler` to `authenticationMux` would be caught at compile time (as opposed to downcasting to `*authenticationMux`).
</blockquote></details>

Done.

---

*[pkg/server/authentication.go, line 333 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD8IF1nFLiQrHVMm4R:-KrU7Kqqjz83v8k_Mdt2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication.go#L333)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

put your name on this TODO
</blockquote></details>

Done.

---

*[pkg/server/authentication.go, line 343 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD8P539RVshko0TAtB:-KrU7RNeVZbH0M5i2i35:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication.go#L343)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

no wrap on this one? you have one in the decode method.
</blockquote></details>

Done.

---

*[pkg/server/authentication_test.go, line 72 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzCqK0XRPpaoSiA5rr:-KrU5NSg2o_f4g5-v23k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L72)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

As above, this comment explains why it is OK to disable authentication, but not why it is helpful to do so.
</blockquote></details>

Done.

---

*[pkg/server/authentication_test.go, line 434 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzDhCk7q1aYvjmevy_:-KrU3d1_2FHlrn2zG_UW:bomenxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication_test.go#L434)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Seems to me that `PostJSONWithHeaders` -> `PostJSONWithResponseor` would be an easy change to make and neatly address the current issues.
</blockquote></details>

There are issues with that, it's weird to return an HttpResponse when the body has already been read and closed.  I think this is fine.

---

*[pkg/server/authentication_test.go, line 536 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD6vs248d6_J3B3VqO:-KrU3cRM_twGrQZnp4IL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication_test.go#L536)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

errors.Errorf, if you don't mind.
</blockquote></details>

Done.

---

*[pkg/server/authentication_test.go, line 562 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD5zDUvHhJjbLgH_ly:-KrU3TTO1bXZQlrst7zh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication_test.go#L562)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

instead of "test auth for ", use a key=value pair, as some tools can use this as metadata:

```
t.Run(fmt.Sprintf("path=%s", tc.path), ...
```
</blockquote></details>

Done.

---

*[pkg/server/server.go, line 981 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD6H6ZboVwTiQ01a-T:-KrEsQGLaHkwEIekgN9Q:buxl6kg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/server.go#L981)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/gwMux/authHandler/
</blockquote></details>

We still expose gwMux directly for the authentication and `/health` endpoints (UserLogin).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/server/authentication.go, line 318 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzBGxiYwYReQvcIgmc:-KrVQvkLI9PHxUKsISI_:bm5fw57) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication.go#L318)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

It's not an internal error, it's an error with the request.
</blockquote></details>

Back to the original point: you're losing the original error completely (it doesn't go to the client, nor the log)

---

*[pkg/server/authentication_test.go, line 536 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD6vs248d6_J3B3VqO:-KrVQe23E7hY-julDL1A:b-8nyi3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication_test.go#L536)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Not done

---

*[pkg/server/server.go, line 981 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD6H6ZboVwTiQ01a-T:-KrVRdNGSym5JK4Z0u4w:b-ndlo0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/server.go#L981)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

We still expose gwMux directly for the authentication and `/health` endpoints (UserLogin).
</blockquote></details>

What does that have to do with it? 
```
authHandler=newAuthMux(s.auth,authHandler)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/server/server.go, line 981 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD6H6ZboVwTiQ01a-T:-KrWSOxneOmkOXGgCOs-:b-b5e9vn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/server.go#L981)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

What does that have to do with it? 
```
authHandler=newAuthMux(s.auth,authHandler)
```
</blockquote></details>

Sorry for the brevity of the previous comment - was typing on phone. What I was saying is that this code can be spelled more as a composition by constructing `authHandler` using the previous value of `authHandler`, rather than referencing `gwMux` again.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 11 of 15 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/server/authentication.go, line 318 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzBGxiYwYReQvcIgmc:-KrWsgx79488mesKNo_4:bjb70mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication.go#L318)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Back to the original point: you're losing the original error completely (it doesn't go to the client, nor the log)
</blockquote></details>

User errors.Wrap, done.

---

*[pkg/server/authentication_test.go, line 536 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD6vs248d6_J3B3VqO:-KrWsHkVg6i8NUpyStuw:b3k5z0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/authentication_test.go#L536)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Not done
</blockquote></details>

Actually done.

---

*[pkg/server/server.go, line 981 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KrD6H6ZboVwTiQ01a-T:-KrWsNBSI-ObwLafYHi4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d971534a3002ade1c0319efed378e510b97b0b0b/pkg/server/server.go#L981)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Sorry for the brevity of the previous comment - was typing on phone. What I was saying is that this code can be spelled more as a composition by constructing `authHandler` using the previous value of `authHandler`, rather than referencing `gwMux` again.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/server/authentication.go, line 318 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzBGxiYwYReQvcIgmc:-KrWxCZxL3HgtLFh90ba:bm4vxh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication.go#L318)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

User errors.Wrap, done.
</blockquote></details>

Looks OK to me, but I'm surprised you're OK with returning this to the client.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/server/authentication.go, line 318 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-KqzBGxiYwYReQvcIgmc:-KrX-idO4CmFaEQIUIq5:b-wo1ih3) ([raw file](https://github.com/cockroachdb/cockroach/blob/349360410f307bb4367c5e23035ac2112eda5364/pkg/server/authentication.go#L318)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Looks OK to me, but I'm surprised you're OK with returning this to the client.
</blockquote></details>

It's fine, it doesn't reveal anything internal to the database itself (just some Go functions) and if they're a third party trying to generate a valid cookie to use with this API, it might help them understand what they're doing wrong.

Is "Looks OK to me" an LGTM?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499)*
<!-- Sent from Reviewable.io -->

tamird commented:
I was hoping you'd address the `PostJSONWithXXX` messiness, but LGTM.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17499#-:-KrX0LZnbTTI33zekQw4:be7zj7g)*
<!-- Sent from Reviewable.io -->
